pkg/core/native/native_test/cryptolib_verification_test.go

```diff
@@ -0,0 +1,924 @@
+		to := util.Uint160{1, 2, 3}
+		amount := 5
+		tx := gasInvoker.PrepareInvokeNoSign(t, "transfer", from, to, amount, nil)
+		tx.Signers = []transaction.Signer{
+			{
+				Account: from,
+				Scopes:  transaction.CalledByEntry,
+			},
+		}
+		neotest.AddNetworkFee(t, e.Chain, tx)
+		neotest.AddSystemFee(e.Chain, tx, -1)
+
+		// Add some more network fee to pay for the witness verification. This value may be calculated precisely,
+		// but let's keep some inaccurate value for the test.
+		tx.NetworkFee += 900_0000
+
+		// This transaction (along with the network magic) should be signed by the user's Koblitz private key.
+		msg := constructMsg(t, uint32(e.Chain.GetConfig().Magic), tx)
+
+		// The users have to sign the hash of the message by their Koblitz key. Collect m signatures from first m keys.
+		// Signatures must be sorted by public key.
+		sigs := make([][]byte, m)
+		for i := range sigs {
+			j := i
+			if i > 0 {
+				j++ // Add some shift to ensure that verification script works correctly.
+			}
+			if i > 3 {
+				j++ // Add more shift for large number of public keys for the same purpose.
+			}
+			sigs[i] = pks[j].SignHash(hash.Keccak256(msg))
+		}
+
+		// Build invocation witness script for the signatures.
+		invBytes := buildKoblitzInvocationScript(t, sigs)
+
+		// Construct witness for signer #0 (the multisig account itself).
+		tx.Scripts = []transaction.Witness{
+			{
+				InvocationScript:   invBytes,
+				VerificationScript: vrfBytes,
+			},
+		}
+
+		// Add transaction to the chain. No error is expected on new block addition. Note, that this line performs
+		// all those checks that are executed during transaction acceptance in the real network.
+		e.AddNewBlock(t, tx)
+
+		// Double-check: ensure funds have been transferred.
+		e.CheckGASBalance(t, to, big.NewInt(int64(amount)))
+	}
+
+	// The proposed multisig verification script.
+	// (279 bytes, 8387820 GAS including Invocation script execution for 3/4 multisig).
+	// The user has to sign the keccak256([4-bytes-network-magic-LE, txHash-bytes-BE]).
+	check(t, buildKoblitzMultisigVerificationScript, constructMessage)
+}
+
+// buildKoblitzMultisigVerificationScript builds witness verification script for m signatures out of n Koblitz public keys.
+// Public keys must be sorted. Signatures (pushed by witness Invocation script) must be sorted by public keys.
+// It checks m out of n multisignature of the following message:
+//
+//	keccak256([4-bytes-network-magic-LE, txHash-bytes-BE])
+func buildKoblitzMultisigVerificationScript(t *testing.T, m int, pubs keys.PublicKeys) []byte {
+	if m > len(pubs) {
+		t.Fatalf("m must be not greater than the number of public keys")
+	}
+
+	criptoLibH := state.CreateNativeContractHash(nativenames.CryptoLib)
+	sort.Sort(pubs)
+
+	// In fact, the following algorithm is implemented via NeoVM instructions:
+	//
+	// func Check(sigs []interop.Signature) bool {
+	// 	var m int = 3
+	// 	if m != len(sigs) {
```

Comment:
Minor, but likely nothing will break if we're to omit this condition (`return numOK == m` should be sufficient).

Reply:
If we skip it, then we have to iterate over sigs from `0` to `len(sigs)` to avoid runtime panic. And in this case user may provide 4 signatures and verification will pass if 3 of them are correct.

Reply:
4/4 is not a bad case for 3/4 requirement. But OK, let's have it as is.